### PR TITLE
perf(graph): improve image scaling quality (issue #39)

### LINKED ISSUE #39: Improve bitmap scaling quality
The currently used image scaling algorithm is nearest neighbor interpolation, and the scaled image is jagged and needs improvement.

### PR DESCRIPTION
**Pull request by ReinForce-II**
Fixes #39 .

Changes proposed in this pull request:
- use bilinear resampling on background images
- add test case

@lc-soft

### CODE REVIEW

**Comment by ReinForce-II** [comment]
Added new function instead  of replacing existing zoom method.

Zoom time tested, cost 8x time than nearest resampling.

File: src/graph.c [hunk 1]
```diff
@@ -1023,6 +1023,155 @@ int Graph_Zoom(const LCUI_Graph *graph, LCUI_Graph *buff, LCUI_BOOL keep_scale,
...
+				d = ARGB(
+				    0xff,
+				    graph->bytes[(index + graph->width + 1) *
+						     graph->bytes_per_pixel +
+						 2],
+				    graph->bytes[(index + graph->width + 1) *
+						     graph->bytes_per_pixel +
+						 1],
+				    graph->bytes[(index + graph->width + 1) *
+						     graph->bytes_per_pixel +
```
[the remark this answers is not in the document]
Comment (ReinForce-II) [reply src/graph.c]: i believe the compiler is able to optimize them out.
if readability is affected, i will cleanup these code in next commit.

File: src/graph.c [hunk 2]
```diff
@@ -1023,6 +1023,155 @@ int Graph_Zoom(const LCUI_Graph *graph, LCUI_Graph *buff, LCUI_BOOL keep_scale,
...
+				rect.x + x;
+			offset =
+			    i * buff->bytes_per_row / buff->bytes_per_pixel + j;
+			if (graph->color_type == LCUI_COLOR_TYPE_ARGB) {
+				a = graph->argb[index];
+				b = graph->argb[index + 1];
+				c = graph->argb[index + graph->width];
+				d = graph->argb[index + graph->width + 1];
+			} else {
+				a = ARGB(
```
Comment (lc-soft) [inline src/graph.c]: The code here is too much, not easy to read, you should simplify the code.

File: src/graph.c [hunk 3]
```diff
@@ -1023,6 +1023,155 @@ int Graph_Zoom(const LCUI_Graph *graph, LCUI_Graph *buff, LCUI_BOOL keep_scale,
 	return 0;
 }
 
+int Graph_Zoom_Bilinear(const LCUI_Graph *graph, LCUI_Graph *buff,
```
Comment (lc-soft) [inline src/graph.c]: You should add links to some reference documents to make it easier for others to understand.

File: configure.ac [hunk 4]
```diff
@@ -226,5 +235,6 @@ echo -e "Build with font-engine support ..... : $font_engine_name"
 echo -e "Build with fontconfig support ...... : $want_fontconfig"
 echo -e "Build with thread support .......... : $thread_name"
 echo -e "Build with video support ........... : $video_driver_name"
+echo -e "Build with optimized scaling ....... : $enable_bilinear_scaling"
```
Comment (lc-soft) [inline configure.ac]: No changes are needed here, you just need to add Graph_Zoom_Bilinear(), I will make some changes in the future to get it called at the right time.

File: test/test_image_scaling.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
+#include <stdlib.h>
```
Comment (lc-soft) [inline test/test_image_scaling.c]: You should add speed test, for example: 

**file**: test_image_scaling_benchmark.c 
**output**:

```
method\image size                      1280x800          1920x1080     4096×2160
Graph_Zoom()                           100ms             200ms         380ms
Graph_ZoomBilinear()                   160ms             320ms         500ms
```

**Comment by ReinForce-II** [comment]
updated code according to requirement; slightly improved the performance (to 6x cost than nearest resampling)

File: test/test_image_scaling_bench.c [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		return -2;
+	}
+	t_color.value = 0xffaa5500;
+	Graph_FillRect(&test_image_scaling, t_color, NULL, false);
+	LOG("%-20s%-20s%s\n", "image size\\method", "Graph_Zoom()",
+	    "Graph_ZoomBilinear()");
+	for (i = 0; i < sizeof(resx) / sizeof(int); i++) {
+		resy = resx[i] * 9 / 16;
+		t0 = LCUI_GetTime();
+		Graph_Init(&test_image_scaling_result);
```
Comment (lc-soft) [inline test/test_image_scaling_bench.c]: These identifiers are named too long, please simplify the naming

- test_image_scaling_result
- test_image_scaling

File: test/test_image_scaling_bench.c [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		Graph_Zoom(&test_image_scaling, &test_image_scaling_result,
+			   false, resx[i], resy);
+		Graph_Free(&test_image_scaling_result);
+		t1 = LCUI_GetTime();
```
Comment (lc-soft) [inline test/test_image_scaling_bench.c]: The return value type of `LCUI_GetTime()` is `int64_t`, please use `int64_t` to declare `t1` and `t2`.

File: test/test_image_scaling_bench.c [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		Graph_Init(&test_image_scaling_result);
+		Graph_ZoomBilinear(&test_image_scaling,
+				   &test_image_scaling_result, false, resx[i],
+				   resy);
+		Graph_Free(&test_image_scaling_result);
+		t2 = LCUI_GetTime();
+		LOG("%dx%-20d%d%-20s%d%s\n", resx[i], resy, t1 - t0, "ms",
```
Comment (lc-soft) [inline test/test_image_scaling_bench.c]: The result is not aligned to the left:

```
./test_image_scaling_bench
image size\method   Graph_Zoom()        Graph_ZoomBilinear()
480x270                 1ms                  28ms
960x540                 6ms                  95ms
1280x720                 13ms                  160ms
1366x768                 16ms                  185ms
1920x1080                24ms                  368ms
2560x1440                42ms                  698ms
3840x2160                109ms                  1457ms
```

Example of improvements that can be referenced:

```
char size_str[32], time1_str[32], time2_str[32];

sprintf(str_size, "%dx%d", resx[i], resy):
sprintf(time1_str, "%dms", t1 - t0);
sprintf(time2_str, "%dms", t1 - t0);
LOG("%-20s%-20s%-20s\n", size_str, time1_str, time2_str);
```

**Comment by ReinForce-II** [comment]
updated code;

**Comment by ReinForce-II** [comment]
resolved

**Comment by lc-soft** [comment]
:+1: 

Thanks

**Pull request merged**